### ISSUE TITLE
Spell: Cannot resist a spell cast on yourself

### ISSUE THREAD

**Issue by joshwhedon** [TrinityCore/TrinityCore#18335]
**Description:** Resistance is ignored when a spell is cast on yourself, even if negative

**Current behaviour:** 
Resistance is not taken into account because of the check for a caster different from the victim in:
```cpp
uint32 Unit::CalcSpellResistance(Unit* victim, SpellSchoolMask schoolMask, SpellInfo const* spellInfo, bool binary) const
```
**Expected behaviour:** Some spells deal damage, and should therefore be resistable by the victim (chilled to the bone from Sindragosa, Shadow Prisons from the blood prince council, etc). We should probably check if the spell is positive instead of checking the caster

**Steps to reproduce the problem:**

1. Get some resistance to shadow
2. Cast 72999
3. No resistance

**Branch(es):** 3.3.5 / master (Tell us which branch(es) this issue affects.)
3.3.5 (that I know of)

**TC rev. hash/commit:**  
https://github.com/TrinityCore/TrinityCore/commit/3cc89c32d551baf8fbc0c4787b79d7e2f7bde2bd

**TDB version:**  
Not relevant

**Operating system:**  
Not relevant

[//]: # (This template is for problem reports. For other types of report, edit it accordingly.)
[//]: # (If this is a crash report, include the crashlog with https://gist.github.com/)
[//]: # (For fixes containing C++, create a Pull Request.)

**Comment by joshwhedon** [comment]
It is unknown however if it is because those spells should be casted with an original caster, and currently are not.

**Comment by MrSmite** [comment]
Are you sure it's not just bad luck? According to wowwiki, spells could have a 0% resistance:

> http://wowwiki.wikia.com/wiki/Resistance
> 
> As of Wrath of the Lich King, the possible percentages of damage resisted were weighted in sections of 0%, 10%, 20%, ..., 100% averaging to the same overall damage mitigation

**Comment by joshwhedon** [comment]
100% certain. I didn't find that out via observation or during the actual fight, some players complained about it, and I don't trust players, so I tested it by debugging. In CalcSpellResistance we return 0 for some spells because they are casted by the victim, and victim can't resist her own spells.
It can be fixed by providing the original caster when casting the spell (I did that for spell proc and for some scripts and it works fine), but not sure it is the proper behaviour anyway.
